### ISSUE TITLE
Unable to use Abstract Class through Nuget with Trigger

### ISSUE THREAD

**Issue by AndrewDessin** [Azure/azure-webjobs-sdk#1624]
I would like to have an abstract class with some basic message processing implementation which then calls other methods to continue processing. 

Example:

        public abstract Task ProcessMessage(BrokeredMessage brokeredMessage, T message);
        
        public abstract Task OnRetriesExpired(T message, Exception ex);
        
        public abstract Task OnParseFailed(BrokeredMessage message, Exception ex);

        public virtual async Task ProcessSubscriptionEvent(
            [ServiceBusTrigger(TopicNameKey, SubscriptionNameKey, AccessRights.Listen)]
            BrokeredMessage brokeredMessage)
        {
            var message = new T();

            // Try Parse Message
            try
            {
                message.Parse(brokeredMessage);
            }
            catch (Exception ex)
            {
                await OnParseFailed(brokeredMessage, ex);
            }

            try
            {
                await ProcessMessage(brokeredMessage, message);
            }
            catch (Exception ex)
            {
                await HandleGlobalExceptionAsync(brokeredMessage, ex, message);
            }
            await brokeredMessage.CompleteAsync();
        }

With an implementation:

        public override async Task ProcessMessage(BrokeredMessage brokeredMessage,  T message)
        {
            //do something
        }

        public override async Task OnRetriesExpired(T message, Exception ex)
        {
            //log
        }

        public override Task OnParseFailed(BrokeredMessage message, Exception ex)
        {
            throw new Exception("Unable to parse message.");
        }

Unfortunately, I am unable to get the trigger to be registered:

`No job functions found. Try making your job classes and methods public. If you're using binding extensions (e.g. ServiceBus, Timers, etc.) make sure you've called the registration method for the extension(s) in your startup code (e.g. config.UseServiceBus(), config.UseTimers(), etc.).`

Unless I add an override implementation for processing:

        public override async Task ProcessSubscriptionEvent(
            [ServiceBusTrigger(TopicNameKey, SubscriptionNameKey, AccessRights.Listen)]
            BrokeredMessage brokeredMessage)
        {
            await base.ProcessSubscriptionEvent(brokeredMessage);
        }

The base class is used via a nuget package.

#### Expected behavior

Trigger is registered and used

#### Actual behavior

Trigger is ignored

#### Known workarounds

Add an override implementation for processing:

        public override async Task ProcessSubscriptionEvent(
            [ServiceBusTrigger(TopicNameKey, SubscriptionNameKey, AccessRights.Listen)]
            BrokeredMessage brokeredMessage)
        {
            await base.ProcessSubscriptionEvent(brokeredMessage);
        }

**Comment by brettsam** [comment]
We may not be looking to base classes for functions. Needs some investigation.

**Comment by paulbatum** [comment]
This behavior is expected based on this code:
https://github.com/Azure/azure-webjobs-sdk/blob/183d8bc77df96a1caed348151046e443375e6a01/src/Microsoft.Azure.WebJobs.Host/Indexers/FunctionIndexer.cs#L27
It looks like this is by design right now. I think its possible that the declared only binding flag could be dropped, but I suspect nobody has thought through whether there are any subtle bugs/issues with this. I recommend you use your workaround of declaring overrides in your derived types that call the base implementation.

**Comment by AndrewDessin** [comment]
Yeah, we'll continue with the workaround for now but, it would be great if we could look into what it would take to remove that flag in the future. 

Thanks!
Andrew
